instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py

```diff
@@ -249,8 +251,8 @@ async def async_response_hook(span, request, response):
 
 class RequestInfo(typing.NamedTuple):
     method: bytes
-    url: URL
-    headers: typing.Optional[Headers]
+    url: httpx.URL
```

Comment:
Yes. I also noticed that the `Headers` could be removed after some typing investigation...

Do you want me to push it? This would be the diff:

<details>
<summary>Details</summary>

diff --git a/instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py b/instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py
index d4e7571f..dfdb9eba 100644
--- a/instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py
+++ b/instrumentation/opentelemetry-instrumentation-httpx/src/opentelemetry/instrumentation/httpx/__init__.py
@@ -237,8 +237,6 @@ from opentelemetry.util.http import remove_url_credentials, sanitize_method
 
 _logger = logging.getLogger(__name__)
 
-URL = typing.Tuple[bytes, bytes, typing.Optional[int], bytes]
-Headers = typing.List[typing.Tuple[bytes, bytes]]
 RequestHook = typing.Callable[[Span, "RequestInfo"], None]
 ResponseHook = typing.Callable[[Span, "RequestInfo", "ResponseInfo"], None]
 AsyncRequestHook = typing.Callable[
@@ -253,17 +251,15 @@ class RequestInfo(typing.NamedTuple):
     method: bytes
     url: httpx.URL
     headers: httpx.Headers | None
-    stream: typing.Optional[
-        typing.Union[httpx.SyncByteStream, httpx.AsyncByteStream]
-    ]
-    extensions: typing.Optional[dict]
+    stream: httpx.SyncByteStream | httpx.AsyncByteStream | None
+    extensions: dict[str, typing.Any] | None
 
 
 class ResponseInfo(typing.NamedTuple):
     status_code: int
     headers: httpx.Headers | None
     stream: typing.Iterable[bytes]
-    extensions: typing.Optional[dict]
+    extensions: dict[str, typing.Any] | None
 
 
 def _get_default_span_name(method: str) -> str:
@@ -274,14 +270,22 @@ def _get_default_span_name(method: str) -> str:
     return method
 
 
-def _prepare_headers(headers: typing.Optional[Headers]) -> httpx.Headers:
+def _prepare_headers(headers: httpx.Headers | None) -> httpx.Headers:
     return httpx.Headers(headers)
 
 
-def _extract_parameters(args, kwargs):
+def _extract_parameters(
+    args: list[typing.Any], kwargs: dict[str, typing.Any]
+) -> tuple[
+    bytes,
+    httpx.URL,
+    httpx.Headers | None,
+    httpx.SyncByteStream | httpx.AsyncByteStream | None,
+    dict[str, typing.Any] | None,
+]:
     if isinstance(args[0], httpx.Request):
         # In httpx >= 0.20.0, handle_request receives a Request object
-        request: httpx.Request = args[0]
+        request = args[0]
         method = request.method.encode()
         url = httpx.URL(remove_url_credentials(str(request.url)))
         headers = request.headers
@@ -300,7 +304,7 @@ def _extract_parameters(args, kwargs):
     return method, url, headers, stream, extensions
 
 
-def _inject_propagation_headers(headers, args, kwargs):
+def _inject_propagation_headers(headers: httpx.Headers | None, args, kwargs):
     _headers = _prepare_headers(headers)
     inject(_headers)
     if isinstance(args[0], httpx.Request):
@@ -311,10 +315,20 @@ def _inject_propagation_headers(headers, args, kwargs):
 
 
 def _extract_response(
-    response: typing.Union[
-        httpx.Response, typing.Tuple[int, Headers, httpx.SyncByteStream, dict]
+    response: httpx.Response
+    | tuple[
+        int,
+        httpx.Headers,
+        httpx.SyncByteStream | httpx.AsyncByteStream,
+        dict[str, typing.Any],
     ],
-) -> typing.Tuple[int, Headers, httpx.SyncByteStream, dict, str]:
+) -> tuple[
+    int,
+    httpx.Headers,
+    httpx.SyncByteStream | httpx.AsyncByteStream,
+    typing.MutableMapping[str, typing.Any],
+    str,
+]:
     if isinstance(response, httpx.Response):
         status_code = response.status_code
         headers = response.headers
@@ -332,7 +346,7 @@ def _extract_response(
 
 def _apply_request_client_attributes_to_span(
     span_attributes: dict,
-    url: typing.Union[str, URL, httpx.URL],
+    url: str | httpx.URL,
     method_original: str,
     semconv: _HTTPStabilityMode,
 ):
@@ -444,7 +458,7 @@ class SyncOpenTelemetryTransport(httpx.BaseTransport):
         *args,
         **kwargs,
     ) -> typing.Union[
-        typing.Tuple[int, "Headers", httpx.SyncByteStream, dict],
+        typing.Tuple[int, httpx.Headers, httpx.SyncByteStream, dict],
         httpx.Response,
     ]:
         """Add request info to span."""
@@ -567,7 +581,7 @@ class AsyncOpenTelemetryTransport(httpx.AsyncBaseTransport):
     async def handle_async_request(
         self, *args, **kwargs
     ) -> typing.Union[
-        typing.Tuple[int, "Headers", httpx.AsyncByteStream, dict],
+        typing.Tuple[int, httpx.Headers, httpx.AsyncByteStream, dict],
         httpx.Response,
     ]:
         """Add request info to span."""


</details>

Reply:
I think we can concentrate everything on https://github.com/open-telemetry/opentelemetry-python-contrib/pull/3098. Does that make sense to you?

Reply:
You mean you want to move this PR to there, or that you want me to just add this last diff there?

> Does that make sense to you?

I'm fine with whatever makes your life easier. :) Let me know what you prefer.

Reply:
Just add this last diff there ✌🏻

Reply:
When this gets in, I'll update that PR. 🙏 

Thanks!